test/e2e/clusterclass_rollout.go

```diff
@@ -590,6 +624,119 @@ func assertMachineDeployments(g Gomega, clusterClassObjects clusterClassObjects,
 	}
 }
 
+func assertMachinePools(g Gomega, clusterClassObjects clusterClassObjects, clusterObjects clusterObjects, cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) {
+	for _, machinePool := range clusterObjects.MachinePools {
+		mpTopology := getMPTopology(cluster, machinePool)
+		mpClass := getMPClass(cluster, clusterClass, machinePool)
+
+		// MachinePool.metadata
+		g.Expect(machinePool.Labels).To(BeEquivalentTo(
+			union(
+				map[string]string{
+					clusterv1.ClusterNameLabel:                    cluster.Name,
+					clusterv1.ClusterTopologyOwnedLabel:           "",
+					clusterv1.ClusterTopologyMachinePoolNameLabel: mpTopology.Name,
+				},
+				mpTopology.Metadata.Labels,
+				mpClass.Template.Metadata.Labels,
+			),
+		))
+		g.Expect(
+			union(
+				machinePool.Annotations,
+			).without(g, clusterv1.RevisionAnnotation),
+		).To(BeEquivalentTo(
+			union(
+				mpTopology.Metadata.Annotations,
+				mpClass.Template.Metadata.Annotations,
+			),
+		))
+
+		// MachinePool.spec.template.metadata
+		g.Expect(machinePool.Spec.Template.Labels).To(BeEquivalentTo(
+			union(
+				map[string]string{
+					clusterv1.ClusterNameLabel:                    cluster.Name,
+					clusterv1.ClusterTopologyOwnedLabel:           "",
+					clusterv1.ClusterTopologyMachinePoolNameLabel: mpTopology.Name,
+				},
+				mpTopology.Metadata.Labels,
+				mpClass.Template.Metadata.Labels,
+			),
+		))
+		g.Expect(machinePool.Spec.Template.Annotations).To(BeEquivalentTo(
+			union(
+				mpTopology.Metadata.Annotations,
+				mpClass.Template.Metadata.Annotations,
+			),
+		))
+
+		// MachinePool InfrastructureMachineTemplate.metadata
+		ccInfrastructureMachineTemplate := clusterClassObjects.InfrastructureMachinePoolTemplateByMachinePoolClass[mpClass.Class]
+		ccInfrastructureMachineTemplateTemplateMetadata := mustMetadata(contract.InfrastructureMachineTemplate().Template().Metadata().Get(ccInfrastructureMachineTemplate))
+		infrastructureMachineTemplate := clusterObjects.InfrastructureMachinePoolTemplateByMachinePool[machinePool.Name]
```

Comment:
infrastructureMachineTemplate should be InfrastructureMachinePool (similar for metadata in l.678)

infrastructureMachineTemplateTemplateMetadata should not exist (there is no template in InfraMachinePool)

Reply:
I'm a bit confused by this one. Isn't there a `DockerMachinePoolTemplate` type and similar ones in the providers? https://github.com/willie-yao/cluster-api/blob/fix-mp-cc-e2e/test/infrastructure/docker/exp/api/v1beta1/dockermachinepooltemplate_types.go#L36

Reply:
Yup there is. Basically the object hierarchies look like this:

ClusterClass:
* MachinePoolClass
  * BootstrapConfigTemplate
  * InfrastructureMachinePoolTemplate

Cluster:
* MachinePool
  * BootstrapConfig
  * InfrastrutureMachinePool

This code here is looking at the Cluster hierarchy